File: java/src/main/java/creational/Builder.java
package creational;

class Meal {

    private final String mainDish;      // required
    private final String side;          // optional
    private final String drink;         // optional
    private final String dessert;       // optional

    // private constructor used only by the builder
    private Meal(Builder b) {
        this.mainDish = b.mainDish;
        this.side = b.side;
        this.drink = b.drink;
        this.dessert = b.dessert;
    }

    // === Builder ===
    public static class Builder {

        private final String mainDish; // required
        private String side;
        private String drink;
        private String dessert;

        public Builder(String mainDish) {
            this.mainDish = mainDish;
        }

        public Builder addSide(String side) {
            this.side = side;
            return this;
        }

        public Builder addDrink(String drink) {
            this.drink = drink;
            return this;
        }

        public Builder addDessert(String dessert) {
            this.dessert = dessert;
            return this;
        }

        public Meal build() {
            return new Meal(this);
        }
    }

    @Override
    public String toString() {
        return "Meal{"
                + "main='" + mainDish + '\''
                + ", side='" + side + '\''
                + ", drink='" + drink + '\''
                + ", dessert='" + dessert + '\''
                + '}';
    }
}

@SuppressWarnings("unused")
final class BuilderExample {

    private BuilderExample() {
    }

    public static void example() {
        Meal meal = new Meal.Builder("Steak")
                .addSide("Salad")
                .addDrink("Tea")
                .build();

        System.out.println(meal);
    }
}

File: rust/src/creational/builder.rs
struct Meal {
    main_dish: String, // required
    side: Option<String>,
    drink: Option<String>,
    dessert: Option<String>,
}

impl Meal {
    fn new(main_dish: String) -> Self {
        Meal {
            main_dish,
            side: None,
            drink: None,
            dessert: None,
        }
    }
}

// === Builder ===
struct MealBuilder {
    meal: Meal,
}

impl MealBuilder {
    fn new(main: &str) -> Self {
        MealBuilder {
            meal: Meal::new(main.to_string()),
        }
    }
    fn add_side(mut self, side: &str) -> Self {
        self.meal.side = Some(side.to_string());
        self
    }
    fn add_drink(mut self, drink: &str) -> Self {
        self.meal.drink = Some(drink.to_string());
        self
    }
    fn add_dessert(mut self, dessert: &str) -> Self {
        self.meal.dessert = Some(dessert.to_string());
        self
    }
    fn build(self) -> Meal {
        self.meal // you could add validation here
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn example() {
        let meal = MealBuilder::new("Steak")
            .add_side("Salad")
            .add_drink("Iced Tea")
            .build();

        println!("Main: {}", meal.main_dish);
    }
}
